### ISSUE TITLE
Implement explicit token genesis declaration as input's unlocking script PreFiX

### ISSUE THREAD

**Issue by A60AB5450353F40E** [bitjson/cashtokens#6]
With some tweaks, this can be adapted from [Group v6.1](https://gitlab.com/0353F40E/group-tokenization/-/blob/e148a4dce1ea3363f7a36e9fdf13ef8f5cb0cfd5/CHIP-2021-02_Unforgeable_Groups_for_Bitcoin_Cash.md#group-genesis)

Rationale: https://bitcoincashresearch.org/t/chip-2022-02-cashtokens-token-primitives-for-bitcoin-cash/725/16

### Input format

* transaction inputs
    * input 0
        * previous output transaction hash, 32 raw bytes
        * previous output index, 4-byte uint
        * unlocking script length, [compact variable length integer](https://reference.cash/protocol/formats/variable-length-integer)
        * unlocking script
            * **OPTIONAL: PFX_TOKEN_GENESIS**, 1-byte constant `0xEF`
                * token genesis hash type, 1-byte
                * token genesis has_nonfungible, 1-byte
                * token genesis commitment length, [compact variable length integer](https://reference.cash/protocol/formats/variable-length-integer)
                * token genesis commitment, 0-40 raw bytes
                * token genesis amount, [compact variable length integer](https://reference.cash/protocol/formats/variable-length-integer)
            * real unlocking script, variable number of raw bytes
        * sequence number, 4-byte uint
    * ...
    * input N

Note: same 0xEF codepoint can be used because it's another context - exclusive to inputs.

### Genesis category ID derivation

`categoryGenID == Hash(categoryGenPreimage)`, where:

- `Hash` is the double SHA-256 hash function, and
- `categoryGenPreimage` is byte concatenation of:
    1. Hash of the input's:
                - prevout transaction hash,
                - prevout index;
    2. Byte concatenation of genesis input parameters:
                - hash type,
                - has_nonfungible,
                - commitment length,
                - commitment,
                - amount;
    3. If `hashType == 0x01` then append the double SHA-256 of the whole prevout UTXO serialization

Note: the input prefix will NOT be part of signature preimage, and doesn't need to be. See notes on [malleability](https://gitlab.com/0353F40E/group-tokenization/-/blob/e148a4dce1ea3363f7a36e9fdf13ef8f5cb0cfd5/CHIP-2021-02_Unforgeable_Groups_for_Bitcoin_Cash.md#group-genesis-malleability) here.

### Token logic validation

Add both the input's prevout token (if present) and the input-generated token to the input side tally of the transaction and enforce token logic over the sums all the same.

### Introspection

In Script VM context, same codepoint 0xEF can be used to encode OP_INPUTTOKENGENESIS which would return concatenation of genesis parameters.

OP_UTXO... introspection opcodes will return the prevout's token if present, NOT the genesis one created with the input.

**Comment by A60AB5450353F40E** [comment]
PS: to better interact with Script, VarInt encoding can be expanded to 8-byte uint when building the preimage, and opcode should return the expanded form

This allows us to keep raw TXes compact, while avoiding having to parse VarInts using Script VM

**Comment by A60AB5450353F40E** [comment]
One more point here...

>Anyway, that was just for some background - it’s about all existing UTXOs having equal “genesis potential” and creators of UTXOs being able to make a covenant that can prohibit, ignore, or require a token genesis on top of itself, with parameters specified by the covenant creator, without having to make sure they put it in the right slot. Also, if genesis is implicit, then you have to do it the roundabout way: verify your genesis output(s) introspected ID against the input’s introspected prevout TXID, if it matches - it’s a genesis. **What if you want to verify the supply being created at genesis? Then you have to tally genesis TX outputs from within Script which is something we’re trying to avoid, no?**

Can't we simply restrict the genesis TX so any new category can have max 1 output in the genesis TX? This way, a Script can simply demand 2 indexes from the spender to verify a genesis setup: one of the input whose TXID will be used, and other of the genesis output.

**Comment by bitjson** [comment]
> Can't we simply restrict the genesis TX so any new category can have max 1 output in the genesis TX? This way, a Script can simply demand 2 indexes from the spender to verify a genesis setup: one of the input whose TXID will be used, and other of the genesis output.

If I understand what you mean – can't such contracts be fooled by a fake preimage? The category ID comes from the spent parent transaction, so the only way a contract can be sure about some genesis information is if the "genesis UTXO" (TXID/0) was a covenant which guarantees some genesis property. And if we wanted to do that, the [covenants-as-standards option here](https://bitcoincashresearch.org/t/chip-2022-02-cashtokens-token-primitives-for-bitcoin-cash/725/21?u=bitjson) is almost certainly more efficient than any sort of preimage reconstruction (36 bytes vs hundreds of bytes).

**Comment by A60AB5450353F40E** [comment]
>If I understand what you mean – can't such contracts be fooled by a fake preimage? The category ID comes from the spent parent transaction, so the only way a contract can be sure about some genesis information is if the "genesis UTXO" (TXID/0) was a covenant which guarantees some genesis property.

Yes. And if this pre-genesis covenant required the genesis to limit the capability of NFTs to mutable and total amount of FTs to some X, it'd have to be coded to verify a fixed number of outputs and limit the total number of outputs in the genesis TX so the spender can't hide some.

>And if we wanted to do that, the [covenants-as-standards option here](https://bitcoincashresearch.org/t/chip-2022-02-cashtokens-token-primitives-for-bitcoin-cash/725/21?u=bitjson) is almost certainly more efficient than any sort of preimage reconstruction (36 bytes vs hundreds of bytes).

Let's further clarify the high-level model of that approach:

1. If we "borrow" some other categoryID to "vet" our genesis setup, does it imply that some blockchain external agent that is the owner of that output must inspect our proposed setup and then decide whether to sign the TX or not based on whether we respect the standard?

2. If the standard-vetting category is set up as a public covenant then a blockchain native agent (an UTXO) could do the inspection and allow itself to be referenced by our TX only if our TX satisfies some standard setup.

Solution 1. requires either a human or some server with which token creators would have to interact to finish signing the TX.

Solution 2. requires none of that, but to spend the standard-vetter covenant you must prove to it that you're setting up the new category as required.

Assuming it is 2., then the standard-vetter covenant must inspect the setup of the genesis TX, right? And if it can have unlimited number of new category's outputs, so the vetter covenant must account for that and limit the total number of outputs and then it must check each one to verify some new category's properties.

If instead it is consensus which guarantees exactly 1 output of a new category in any genesis TX, then the vetter contract only needs to check 1 output, index provided by the spender, and it will know everything about the genesis setup. There can be any number of other outputs, even multiple categories can be created in a single TX, you could tie a batch of categories like this. This is the advantage of having a single root of any category's sub-DAG, it's enough to inspect that 1 node and you know everything about possible evolution of the category.

The input PFX moves that root node from an output to an input. We can stick it to the input while keeping the txid/0 convenience.

This would let contracts verify genesis setup of a new category by using a single input introspection opcode, while avoiding boxing transactions to a particular format. Any number of genesis operations could mingle with any number of inputs and outputs, without affecting ability of UTXOs to verify those or requiring TXes to be built in a particular way so contracts can validate them.

Plus, it would open a clean upgrade path to other genesis types:

1. generate ID from previous block hash when placed on coinbase inputs and signal using hash_type=0x01
2. generate from other prevout indexes by H(txid, index) and signal using hash_type=0x02
3. custom structure as previously proposed and signal using hash_type=0x03...

## Genesis Input Take 2

### Input format

* transaction inputs
    * input 0
        * previous output transaction hash, 32 raw bytes
        * previous output index, 4-byte uint
        * unlocking script length, [compact variable length integer](https://reference.cash/protocol/formats/variable-length-integer)
        * unlocking script
            * **OPTIONAL: PFX_TOKEN_GENESIS**, 1-byte constant `0xEF`
                * token genesis hash type, 1-byte
                * token genesis output type, 1-byte
                * OPTIONAL: token genesis commitment length, [compact variable length integer](https://reference.cash/protocol/formats/variable-length-integer)
                * OPTIONAL: token genesis commitment, variable number of raw bytes
                * OPTIONAL: token genesis amount, [compact variable length integer](https://reference.cash/protocol/formats/variable-length-integer)
            * real unlocking script, variable number of raw bytes
        * sequence number, 4-byte uint
    * ...
    * input N

Note: same 0xEF codepoint can be used because it's another context - exclusive to inputs.

### Genesis category ID validation

    If (input.category_genesis.hash_type == 0x00) {
        if (input.prevout.index != 0x00)
            fail();
        else
             input.category_genesis.id = input.prevout.txid;
    }
    else
        fail();

### Token logic validation

Add both the input's prevout token (if present) and the input-generated token (input.category_genesis.id, input.category_genesis.type, input.category_genesis.amount, input.category_genesis.commitment) to the input side tally of the transaction and enforce token logic over the sums all the same.

### Introspection

In Script VM context, same codepoint 0xEF can be used to encode OP_INPUTTOKENGENESIS which would return concatenation of genesis parameters, with VarInts expanded to 8-bytes for easier manipulation by Script.

OP_UTXO... introspection opcodes will return the prevout's token if present, NOT the genesis one created with the input.

**Comment by A60AB5450353F40E** [comment]
ok, even if we don't have this and only go with implicit `category_ID=TXID` from `prevout.index==0` inputs we can still upgrade in the future if it is demonstrated it would solve a problem... we don't need it now, I'm removing the explicit genesis input nanotation from GroupV8 (now called "unforgeable categories") https://gitlab.com/0353F40E/group-tokenization/-/commit/3f5c77cc90769859cc1e2d37a5e21e1369cfcf96
